src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/ManagedIndexCoordinator.kt

```diff
@@ -0,0 +1,338 @@
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.OpenForTesting
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.createManagedIndexRequest
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.deleteManagedIndexRequest
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.getCreateManagedIndexRequests
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.getDeleteManagedIndexRequests
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.getSweptManagedIndexSearchRequest
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.getUpdateManagedIndexRequests
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.util.updateManagedIndexRequest
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.action.DocWriteRequest
+import org.elasticsearch.action.bulk.BulkRequest
+import org.elasticsearch.action.bulk.BulkResponse
+import org.elasticsearch.action.search.SearchResponse
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.ClusterChangedEvent
+import org.elasticsearch.cluster.ClusterState
+import org.elasticsearch.cluster.ClusterStateListener
+import org.elasticsearch.cluster.LocalNodeMasterListener
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.common.bytes.BytesReference
+import org.elasticsearch.common.component.LifecycleListener
+import org.elasticsearch.common.settings.Settings
+import org.elasticsearch.common.unit.TimeValue
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler
+import org.elasticsearch.common.xcontent.NamedXContentRegistry
+import org.elasticsearch.common.xcontent.XContentHelper
+import org.elasticsearch.common.xcontent.XContentParser
+import org.elasticsearch.common.xcontent.XContentType
+import org.elasticsearch.threadpool.Scheduler
+import org.elasticsearch.threadpool.ThreadPool
+
+/**
+ * Listens for cluster changes to pick up new indices to manage.
+ * Sweeps the cluster state and [INDEX_STATE_MANAGEMENT_INDEX] for [ManagedIndexConfig].
+ *
+ * This class listens for [ClusterChangedEvent] to pick up on [ManagedIndexConfig] to create, update, or delete.
+ * Also sets up a background process that sweeps the cluster state for [ClusterStateManagedIndexConfig]
+ * and the [INDEX_STATE_MANAGEMENT_INDEX] for [SweptManagedIndexConfig]. It will then compare these
+ * ManagedIndices to appropriately create, delete, or update each [ManagedIndexConfig]. Each node that has
+ * the [IndexStateManagementPlugin] installed will have an instance of this class, but only the elected
+ * master node will set up the background sweep process and listen for [ClusterChangedEvent].
+ */
+@Suppress("TooManyFunctions")
```

Comment:
How many function is considered `TooManyFunctions` here? Is this something specific to Kotlin or the linter we use?

Reply:
Specific to detekt, we can change the rules in the config, but I think having the default is fine. It's just this specific class implements a few different listeners so the function count went up quite a bit with things like onMaster, offMaster, afterStart, beforeStop, enable, disable, etc.